File: ruby/submitorder.rb
require './oneflowclient'
require './models/order'
require './models/item'
require './models/component'
require './models/shipment'
require './models/carrier'
require './models/address'
require 'json'

#initialise the sdk
endpoint = "ONEFLOW-ENDPOINT" # Don't forget the /api at the end of the endpoint
token = "ONEFLOW-TOKEN"
secret = "ONEFLOW-SECRET"

#order settings
destination = "DESTINATION-ACCOUNT"
orderId = "order-"+rand(1000000).to_s
itemId = orderId + "-1"
fetchUrl = "FETCHURL"
quantity = 1
skuCode = "SKU-CODE"
componentCode = "COMPONENT-CODE"
carrierCode = "CARRIER-CODE"
carrierService = "CARRIER-SERVICE"
name = "Nigel Watson"
address1 = "999 Letsbe Avenue"
town = "London"
postcode = "EC2N 0BJ"
isoCountry = "GB"

# Retry options
class Options
    attr_accessor :retries

    def initialize()
        @retries = 3
    end

    def retryCondition(response)
        return response.code != 200 # retry everything but OK
    end

    def retryDelay(response, retryCount)
        sleep 1 # just delay 1 sec
    end
end

#create the onelflow client. Pass Options.new as 4th argument and customize logic or skip for defaults (exponential retry in case of 429)
client = OneflowClient.new(endpoint, token, secret)

# No need to edit below here

order = client.create_order(destination)
order.orderData.sourceOrderId = orderId
item = Item.new
item.sku = skuCode
item.quantity = quantity
item.sourceItemId = itemId

component = Component.new
component.code = componentCode
component.fetch = false
component.path = fetchUrl
item.components.push(component)
order.orderData.items.push(item)

shipment = Shipment.new
shipment.carrier = Carrier.new(carrierCode, carrierService)
address = Address.new
address.name = name
address.address1 = address1
address.town = town
address.isoCountry = isoCountry
address.postcode = postcode
shipment.shipTo = address
order.orderData.shipments.push(shipment)

response = client.submit_order
response_json = JSON.parse(response.body)

if (response.code != 200)
    puts "Error"
    puts "====="
    puts response_json["error"] ? response_json["error"]["message"] : response_json["message"]
    if (response_json["error"] && response_json["error"]["validations"])
        response_json["error"]["validations"].each {|val| puts val["path"] + " -> " + val["message"]}
    end
else
    saved_order = response_json["order"]

    puts "Success"
    puts "======="
    puts "Order ID        : " + saved_order["_id"]

#    for file in saved_order["files"]
#        puts "Uploading File  : " + file["_id"]
#        # res = client.upload_file(file, "files/" + file["path"])
#    end
end

File: python/submit-order.py
#!/usr/bin/python

__author__ = 'oneflow'

import json, string, random, os
from OneflowSDK import OneflowSDK

def id_generator(size=6, chars=string.ascii_uppercase + string.digits):
	return ''.join(random.choice(chars) for _ in range(size))

def validate_order(order):
	print("Validating Order")
	return client.request("POST", '/api/order/validate', order)

def submit_order(order):
	print("Submitting Order")
	return client.request("POST", '/api/order', order)

#oneflow acccess credentials
token = os.environ['ONEFLOW_TOKEN']
secret = os.environ['ONEFLOW_SECRET']
endpoint = 'http://stage.oneflowcloud.com'

#OneflowSDK instance
client = OneflowSDK(endpoint, token, secret)

#Here are the variables required for a single item order
#this is not a full set of fields available
destination = "oneflow"
orderId = id_generator()
itemId = id_generator()
sku = "card"
quantity = 1
fetchPath = "https://files-static.hpsiteflow.com/samples/CardSample.pdf"
componentCode = "card"
shipTo = {
	"name":           "Nigel Watson",
	"address1":       "1 Primrose Street",
	"town":           "London",
	"postcode":       "12345",
	"isoCountry":     "GB",
	"email":          "info@oneflowsystems.com",
	"phone":          "01234567890"
}
carrier = {
	"code":       "royalmail",
	"service":    "firstclass"
}

#nothing to change below here

#create an item - this goes into the order below
item = {
	"sourceItemId":         itemId,
	"sku":                  sku,
	"quantity":             quantity,
	"components": [{
		"code":               componentCode,
		"path":               fetchPath,
		"fetch":			  "true"
	}]
}

#create a shipment - this goes into the order below
shipment = {
	"shipTo":  shipTo,
	"carrier": carrier
}

#put together the complete order
order = {
  "destination": {
    "name":       destination
  },
  "orderData": {
    "sourceOrderId":        orderId,
    "items": [ item ],
    "shipments": [ shipment ]
  }
}

#validate the order
print(validate_order(json.dumps(order)))

#finally, submit the order
print(submit_order(json.dumps(order)))
